### PR TITLE
QUICK-307 Refactor for re-use

### PR DESCRIPTION
**Pull request by mminns**
Hi

This is a slightly speculative PR.

I've been investigating re-using the JPT HWR reporting functionality to report on CPT results and therefore giving a standard set of HWR graphs and results.

This is a potential 1st step. It refactors this project to put the reporting function into the _main_ code rather than _test_.

There are a couple of additional changes to break the hard connections between Jira testing and the reporting.

I have another branch that makes use of this refactored code to create a basic command line tool for downloading and generating reports from existing JPT results.

It is possible it makes more sense to extra the HWR code into a new module **report-hardware** ?

So feel free to say no to this PR :)

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -65,15 +65,16 @@ task<Test>("cleanUpAfterBamboo").apply {
...
-    testCompile("com.atlassian.performance.tools:aws-infrastructure:[2.15.0,3.0.0)")
-    testCompile("com.atlassian.performance.tools:aws-resources:[1.3.4,2.0.0)")
-    testCompile("com.atlassian.performance.tools:concurrency:[1.0.0,2.0.0)")
-    testCompile("org.apache.commons:commons-csv:1.4")
+    compile("com.atlassian.performance.tools:aws-infrastructure:[2.15.0,3.0.0)")
+    compile("com.atlassian.performance.tools:aws-resources:[1.3.4,2.0.0)")
+    compile("com.atlassian.performance.tools:concurrency:[1.0.0,2.0.0)")
+    compile("org.apache.commons:commons-csv:1.4")
+    compile( "com.xenomachina:kotlin-argparser:2.0.7")
     testCompile("org.eclipse.jgit:org.eclipse.jgit:4.11.0.201803080745-r")
```
Comment (mminns) [inline build.gradle.kts:78]: Change the compliation target to be _main_ not _test_

File: gradle/dependency-locks/testRuntime.lockfile [hunk 2]
```diff
@@ -39,11 +40,14 @@ com.googlecode.javaewah:JavaEWAH:1.1.6
...
 com.jcraft:jsch:0.1.54
 com.jcraft:jzlib:1.1.3
+com.numericalmethod:suanshu-20120606:1.0.1-atlassian-hosted
 com.squareup.okhttp3:okhttp:3.9.1
 com.squareup.okio:okio:1.13.0
 com.sun.istack:istack-commons-runtime:3.0.5
 com.sun.xml.fastinfoset:FastInfoset:1.2.13
 com.typesafe:config:1.2.1
+com.xenomachina:kotlin-argparser:2.0.7
+com.xenomachina:xenocom:0.0.7
```
Comment (mminns) [inline gradle/dependency-locks/testRuntime.lockfile:50]: These are a couple of hangovers from my command line tool branch.

File: src/main/kotlin/com/atlassian/performance/tools/hardware/HardwareExistingResults.kt [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import com.atlassian.performance.tools.report.api.result.EdibleResult
+import com.atlassian.performance.tools.virtualusers.api.TemporalRate
+import com.atlassian.performance.tools.workspace.api.TaskWorkspace
+import com.atlassian.performance.tools.workspace.api.TestWorkspace
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+import java.io.File
+import java.time.Duration
+
+class HardwareExistingResults(
```
Comment (mminns) [inline src/main/kotlin/com/atlassian/performance/tools/hardware/HardwareExistingResults.kt:24]: Refactor the code for handling existing results out in to a new class

File: src/main/kotlin/com/atlassian/performance/tools/hardware/HardwareExploration.kt [hunk 4]
```diff
@@ -204,50 +204,11 @@ class HardwareExploration(
         hardware: Hardware,
         executor: ExecutorService
     ): HardwareTestResult {
-        val reusableResults = reuseResults(hardware)
+        val reusableResults = existingResults.reuseResults(hardware)
         val missingResultCount = repeats - reusableResults.size
         val freshResults = runFreshResults(hardware, missingResultCount, executor)
         val allResults = reusableResults + freshResults
-        return coalesce(allResults, hardware)
```
Comment (mminns) [inline src/main/kotlin/com/atlassian/performance/tools/hardware/HardwareExploration.kt:211]: Code has been moved to _HardwareExistingResults_

File: src/main/kotlin/com/atlassian/performance/tools/hardware/HardwareMetric.kt [hunk 5]
```diff
@@ -38,7 +38,7 @@ class HardwareMetric(
         val throughput = AccessLogThroughput().gauge(results.results.toFile())
         val overallError = ErrorGauge().measureOverall(metrics)
         val maxActionError = ErrorGauge().measureMaxAction(metrics)
-        return HardwareTestResult(
+            return HardwareTestResult(
```
Comment (mminns) [inline src/main/kotlin/com/atlassian/performance/tools/hardware/HardwareMetric.kt:41]: extra space

File: src/main/kotlin/com/atlassian/performance/tools/hardware/HardwareRecommendationEngine.kt [hunk 6]
```diff
@@ -45,66 +46,23 @@ class HardwareRecommendationEngine(
...
-            val dbRecommendationTable = tabularize(dbRecommendations)
-            return ReportedRecommendations(
-                description = scale.description,
-                recommendations = dbRecommendations,
-                reports = listOfNotNull(jiraReport, dbRecommendationChart, dbRecommendationTable)
-                    + jiraExploration.reports
-                    + dbExploration.reports
-            )
+
+            return HardwareReportEngine().reportedRecommendations(scale.description,
```
Comment (mminns) [inline src/main/kotlin/com/atlassian/performance/tools/hardware/HardwareRecommendationEngine.kt:53]: Refactored into a new class

File: src/main/kotlin/com/atlassian/performance/tools/hardware/HardwareReportEngine.kt [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+import com.atlassian.performance.tools.hardware.ReportedExploration
+import com.atlassian.performance.tools.hardware.ReportedRecommendations
+import com.atlassian.performance.tools.hardware.report.*
+import com.atlassian.performance.tools.hardware.report.HardwareExplorationChart
+import com.atlassian.performance.tools.jvmtasks.api.TaskTimer
+import com.atlassian.performance.tools.workspace.api.git.GitRepo
+import java.io.File
+import java.nio.file.Path
+
+class HardwareReportEngine {
```
Comment (mminns) [inline src/main/kotlin/com/atlassian/performance/tools/hardware/HardwareReportEngine.kt:14]: Refactored report code into a new class

File: src/main/kotlin/com/atlassian/performance/tools/hardware/OutcomeRequirements.kt [hunk 8]
```diff
@@ -15,4 +16,16 @@ class OutcomeRequirements(
             && (result.overallError.ratio < overallErrorThreshold.ratio)
             && (result.maxActionError!!.ratio < maxActionErrorThreshold)
     }
+
+    fun areReliable(
```
Comment (mminns) [inline src/main/kotlin/com/atlassian/performance/tools/hardware/OutcomeRequirements.kt:20]: Make the reliability check a requirement

File: src/main/kotlin/com/atlassian/performance/tools/hardware/aws/HardwareRuntime.kt [hunk 9]
```diff
@@ -13,7 +13,7 @@ import java.nio.file.Paths
 import java.time.Duration
 import java.util.*
 
-object IntegrationTestRuntime {
+object HardwareRuntime {
```
Comment (mminns) [inline src/main/kotlin/com/atlassian/performance/tools/hardware/aws/HardwareRuntime.kt:16]: Renamed as its no longer tied to tests.

**Comment by dagguh** [comment]
`src/main` only needs the classes, which are going to become API. Adding API is easy, removing it is hard.
Let's only add the MVP API - the reporting bits. We'll still be able to share more HWR mechanisms for reuse in the future. This should drastically reduce the size of the diff.

**Comment by mminns** [comment]
@dagguh thanks for taking a look. 

This was roughly speaking the minimum to get the exploration chart and csv generating outside of running a test.
Could you give me an example or two of things you feel should **not** be in _main_ ? 
Thanks

**Comment by dagguh** [comment]
> Could you give me an example or two of things you feel should **not** be in _main_ ?

* provisioning, e.g. `WgetOracleJdk`, `ConfigurableMysqlDatabase`, `CopiedDocker`, `BestEffortProfiler`, anything in `lib/infrastructure`, `DatabaseOverride`
* `JiraNodeTuning`, etc.
* benchmark orchestration, e.g. `FailureTolerance`, `DbExplorationGuidance`, 
* S3 caching, e.g. `S3Listing`, `EtagCache`

Some things are tangled up, because I didn't always follow the [Interface Segregation Principle](https://github.com/dwmkerr/hacker-laws/tree/2af637016656686dad1572ecd8d11c9a2ff9d6ca#the-interface-segregation-principle). E.g. the `HardwareMetric` depends on `ApplicationScale` only to extract VU data, but accidentally pulls in `AdminDataset`(provisioning) and `cacheKey` (external caching).

**Pull request closed without merging**